asterixdb/asterix-algebra/src/main/java/org/apache/asterix/optimizer/rules/FilterRefineSpatialDistanceJoin.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.hyracks.algebricks.core.algebra.operators.logical.AbstractBinaryJoinOperator;
+import org.apache.hyracks.algebricks.core.algebra.operators.logical.AbstractLogicalOperator;
+import org.apache.hyracks.algebricks.core.rewriter.base.IAlgebraicRewriteRule;
+
+public class FilterRefineSpatialDistanceJoin implements IAlgebraicRewriteRule {
+
+    private static final int LEFT = 0;
+    private static final int RIGHT = 1;
+
+    @Override
+    public boolean rewritePre(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
+            throws AlgebricksException {
+        return false;
+    }
+
+    @Override
+    public boolean rewritePost(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
+            throws AlgebricksException {
+
+        AbstractLogicalOperator op = (AbstractLogicalOperator) opRef.getValue();
+        if (op.getOperatorTag() != LogicalOperatorTag.INNERJOIN
+                && op.getOperatorTag() != LogicalOperatorTag.LEFTOUTERJOIN) {
+            return false;
+        }
+
+        AbstractBinaryJoinOperator joinOp = (AbstractBinaryJoinOperator) op;
+        Mutable<ILogicalExpression> joinConditionRef = joinOp.getCondition();
+        ILogicalExpression joinCondition = joinConditionRef.getValue();
+
+        if (joinCondition.getExpressionTag() != LogicalExpressionTag.FUNCTION_CALL) {
+            return false;
+        }
+
+        AbstractFunctionCallExpression funcExpr = (AbstractFunctionCallExpression) joinCondition;
+        if (!(funcExpr.getFunctionIdentifier().equals(BuiltinFunctions.LT)
+                || funcExpr.getFunctionIdentifier().equals(BuiltinFunctions.LE)
+                || funcExpr.getFunctionIdentifier().equals(BuiltinFunctions.EQ))) {
+            return false;
+        }
+
+        List<Mutable<ILogicalExpression>> inputExprs = funcExpr.getArguments();
+        if (inputExprs.size() != 2) {
+            return false;
+        }
+
+        ILogicalExpression leftOperatingExpr = inputExprs.get(LEFT).getValue();
+        ILogicalExpression rightOperatingExpr = inputExprs.get(RIGHT).getValue();
+
+        if (!(((ScalarFunctionCallExpression) leftOperatingExpr).getFunctionIdentifier()
+                .equals(BuiltinFunctions.ST_DISTANCE))) {
+            return false;
+        }
+
+        LogicalVariable inputVar0;
+        LogicalVariable inputVar1;
+        IAlgebricksConstantValue distanceVar;
+
+        inputVar0 = ((VariableReferenceExpression) ((ScalarFunctionCallExpression) leftOperatingExpr).getArguments()
+                .get(LEFT).getValue()).getVariableReference();
+        inputVar1 = ((VariableReferenceExpression) ((ScalarFunctionCallExpression) leftOperatingExpr).getArguments()
+                .get(RIGHT).getValue()).getVariableReference();
+        distanceVar = ((ConstantExpression) rightOperatingExpr).getValue();
+
+        ScalarFunctionCallExpression enlargedLeft = new ScalarFunctionCallExpression(
```

Comment:
Is this information only needed by the physical spatial join operator? If so, I think you can delay adding it until you pick the spatial physical operator.

Reply:
Yes, you are right.
@akilsevim : could you please check this?